File: appendixA/src/main/csharp/Movie/DiscountCondition.cs
using System;

namespace MovieReservation
{
    public interface DiscountCondition
    {
        bool IsSatisfiedBy(Screening screening);
    }

    public class SequenceCondition : DiscountCondition
	{
    	private int sequence;

    	public SequenceCondition(int sequence)
    	{
    		this.sequence = sequence;
    	}

    	public bool IsSatisfiedBy(Screening screening)
    	{
    		return screening.IsSequence(sequence);
    	}
    }

    public class PeriodCondition : DiscountCondition
    {
    	private DayOfWeek dayOfWeek;
    	private TimeSpan startTime;
    	private TimeSpan endTime;

        public PeriodCondition(DayOfWeek dayOfWeek, TimeSpan startTime, TimeSpan endTime)
    	{
    		this.dayOfWeek = dayOfWeek;
    		this.startTime = startTime;
    		this.endTime = endTime;
    	}

    	public bool IsSatisfiedBy(Screening screening)
    	{
    		return screening.StartTime.DayOfWeek.Equals(dayOfWeek) &&
                startTime <= screening.StartTime.TimeOfDay &&
		        endTime >= screening.StartTime.TimeOfDay;
    	}
    }
}

File: chapter04/src/main/java/org/eternity/movie/step02/DiscountCondition.java
package org.eternity.movie.step02;

import java.time.DayOfWeek;
import java.time.LocalTime;

public class DiscountCondition {
    private DiscountConditionType type;

    private int sequence;

    private DayOfWeek dayOfWeek;
    private LocalTime startTime;
    private LocalTime endTime;

    public DiscountCondition(int sequence){
        this.type = DiscountConditionType.SEQUENCE;
        this.sequence = sequence;
    }

    public DiscountCondition(DayOfWeek dayOfWeek, LocalTime startTime, LocalTime endTime){
        this.type = DiscountConditionType.PERIOD;
        this.dayOfWeek= dayOfWeek;
        this.startTime = startTime;
        this.endTime = endTime;
    }

    public DiscountConditionType getType() {
        return type;
    }

    public boolean isDiscountable(DayOfWeek dayOfWeek, LocalTime time) {
        if (type != DiscountConditionType.PERIOD) {
            throw new IllegalArgumentException();
        }

        return this.dayOfWeek.equals(dayOfWeek) &&
                this.startTime.compareTo(time) <= 0 &&
                this.endTime.compareTo(time) >= 0;
    }

    public boolean isDiscountable(int sequence) {
        if (type != DiscountConditionType.SEQUENCE) {
            throw new IllegalArgumentException();
        }

        return this.sequence == sequence;
    }
}
